source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp

```diff
@@ -523,11 +526,32 @@ inline void BaseControllerInterface::set_parameter_value(const std::string& name
 
 template<typename T>
 inline void BaseControllerInterface::add_input(const std::string& name, const std::string& topic_name) {
-  auto buffer = realtime_tools::RealtimeBuffer<std::shared_ptr<modulo_core::EncodedState>>();
-  auto input = ControllerInput(buffer);
-  create_input(input, name, topic_name);
-  input_message_pairs_.insert_or_assign(
-      name, modulo_core::communication::make_shared_message_pair(std::make_shared<T>(), get_node()->get_clock()));
+  if constexpr (modulo_core::concepts::CustomT<T>) {
+    auto buffer = realtime_tools::RealtimeBuffer<T>();
+    auto input = ControllerInput(buffer);
+    auto parsed_name = validate_and_declare_signal(name, "input", topic_name);
+    if (!parsed_name.empty()) {
+      inputs_.insert_or_assign(parsed_name, input);
+      custom_input_configuration_callables_.insert_or_assign(
+          name, [this](const std::string& name, const std::string& topic) {
+            auto subscription =
+                get_node()->create_subscription<T>(topic, qos_, [this, name](const std::shared_ptr<T> message) {
+                  // TODO: should we try catch bad_any_cast in here?
```

Comment:
We should just be consistent overall

Reply:
In #150, you don't catch the bad_any_cast either, you can decide what you prefer. I wouldn't know how an exception could happen here but might be safer, Idk

Reply:
I don't think it can happen at this point, no. So I'll just remove the comment